[quant][graphmode] Support propagate dequantize for nodes with multiple outputs

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39645 [quant][graphmode] Support quantized::batch_norm2d_relu fusion for tracing
* #39644 [quant][graphmode] Support quantization for `aten::append`
* #39615 [quant][graphmode] Fix a corner case in handling `if` in insert_observers
* **#39551 [quant][graphmode] Support propagate dequantize for nodes with multiple outputs**
* #39550 [quant][graphmode][refactor] propagateQuantizationOps

Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942108](https://our.internmc.facebook.com/intern/diff/D21942108)

jerryzh168 committed:
[quant][graphmode] Support propagate dequantize for nodes with multiple outputs

Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support propagate dequantize for nodes with multiple outputs"


Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support propagate dequantize for nodes with multiple outputs"


Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support propagate dequantize for nodes with multiple outputs"


Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -932,16 +932,19 @@ def __init__(self):
+                return x + y
         x = torch.randn([1, 3, 10, 10], dtype=torch.float)
         x = torch.quantize_per_tensor(x, 0.5, 1, torch.quint8)
+        input = [x, x]
```
Should we also add a test for tracing here?

jerryzh168 replied:
I think we probably need people to expand all tests with tracing

jerryzh168 replied:
this one probably doesn't make much difference, but I can add that

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -932,16 +932,19 @@ def __init__(self):
+                return x + y
         x = torch.randn([1, 3, 10, 10], dtype=torch.float)
         x = torch.quantize_per_tensor(x, 0.5, 1, torch.quint8)
+        input = [x, x]
```
Do both quantize_per_tensor and dequantize now support list of tensors?

jerryzh168 replied:
yeah that's correct: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L3580, but I think we don't really need the list version of quantize_per_tensor right now

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -950,7 +953,7 @@ def forward(self, x):
                    .check("dequantize") \
                    .run(m.graph)
-        res = get_forward(m._c)(x)
+        res = m(input)
```
Wouldnt the listunpack op be needed prior to the maxpool and avgpool even after the swap dequantize pass?

jerryzh168 replied:
yeah, it's still there. I forgot to add the check for that

jerryzh168 committed:
Update on "[quant][graphmode] Support propagate dequantize for nodes with multiple outputs"


Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support propagate dequantize for nodes with multiple outputs"


Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -890,12 +890,35 @@ void propagateQuantizationOps(Block* block) {
         }
       }
     } else {
+      // For this category of ops, we need to
```
What is the mechanism for warning/providing errors for quantization? In this case, it will be good to error out with a specific message.
```
if (n->outputs().size() > 1) {
        // Factoring out dequantize for if blocks with multiple outputs	        // Factoring out dequantize for if blocks with multiple outputs
        // is not supported right now	        // is not supported right now
        continue;	        continue;
      }	      }
```

jerryzh168 replied:
this should probably be a warning since we might not need this functionality when there are multiple outputs, and the result would be incorrect if we do need the support.

jerryzh168 replied:
In general when it's a clear case we don't support, we'll throw an error

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -890,12 +890,35 @@ void propagateQuantizationOps(Block* block) {
         }
       }
     } else {
+      // For this category of ops, we need to
```
Also, when you mean "this category" of ops, do you mean operations that have multiple inputs or outputs?.

jerryzh168 replied:
no, this is for ops that we quantize by propagating dequantize ops, e.g. flatten

vkuzo replied:
> no, this is for ops that we quantize by propagating dequantize ops, e.g. flatten

maybe we can include this in the comment?

jerryzh168 committed:
Update on "[quant][graphmode] Support propagate dequantize for nodes with multiple outputs"


Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942108](https://our.internmc.facebook.com/intern/diff/D21942108)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support propagate dequantize for nodes with multiple outputs"


Summary:
e.g. prim::ListUnpack

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942108](https://our.internmc.facebook.com/intern/diff/D21942108)

[ghstack-poisoned]
